File: src/org/naderica/whyscala/dto/java/Name.java
package org.naderica.whyscala.dto.java;

public class Name {

	public static void main(String[] args) {
		final Name name = new Name("Ali", "Baba");
		System.out.println(name);

		name.setFirstName("Donald");
		name.setLastName("Duck");
		System.out.println(name);
	}

	private String firstName;

	private String lastName;

	public Name(String firstName, String lastName) {
		super();
		this.firstName = firstName;
		this.lastName = lastName;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (getClass() != obj.getClass()) {
			return false;
		}
		Name other = (Name) obj;
		if (this.firstName == null) {
			if (other.firstName != null) {
				return false;
			}
		} else if (!this.firstName.equals(other.firstName)) {
			return false;
		}
		if (this.lastName == null) {
			if (other.lastName != null) {
				return false;
			}
		} else if (!this.lastName.equals(other.lastName)) {
			return false;
		}
		return true;
	}

	public String getFirstName() {
		return this.firstName;
	}

	public String getLastName() {
		return this.lastName;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result
				+ ((this.firstName == null) ? 0 : this.firstName.hashCode());
		result = prime * result
				+ ((this.lastName == null) ? 0 : this.lastName.hashCode());
		return result;
	}

	public void setFirstName(String firstName) {
		this.firstName = firstName;
	}

	public void setLastName(String lastName) {
		this.lastName = lastName;
	}

	@Override
	public String toString() {
		return "Name [firstName=" + this.firstName + ", lastName="
				+ this.lastName + "]";
	}

}

File: src/org/naderica/whyscala/dto/scala/Name.scala
package org.naderica.whyscala.dto.scala

/**
 * Case classes encourage developers to prefer immutable objects.  When <code>val</code> or <code>var</code> is dropped (default behaviour), Scala creates
 * a <code>val</code> reference, which makes the class immutable.
 * 
 * <p>Here, we are showing the usage of <code>var</code> to illustrate how it matches a mutable Domain Transfer Object (DTO) in Java.
 * 
 * <p>For immutable objects, Scala automatically creates both getters and setters.
 */
case class Name(var firstName: String, var lastName: String)

object Name {
  def main(args: Array[String]) {
    val name = new Name("Ali", "Baba")
    println(name)

    name.firstName = "Donald"
    name.lastName = "Duck"
    println(name)

  }
}